### ISSUE TITLE
[ROCm] CUDACachingAllocator: hipMalloc SIGSEGV after ~27 loop iterations on ROCm 7.x (AMD MI300A)

### ISSUE THREAD

**Issue by qiyang-ustc** [pytorch/pytorch#185826]
## 🐛 Bug

On ROCm 7.x (tested: PyTorch 2.12.0+rocm7.2, AMD Instinct MI300A), running a tight loop of 5-tensor `torch.einsum` contractions causes a **silent SIGSEGV (exit 139)** after approximately 27 iterations. The crash occurs inside `hipMalloc` in the HIP VA allocator — PyTorch's GC retry loop never fires because `hipMalloc` kills the process rather than returning `hipErrorOutOfMemory`.

### Environment

| | |
|---|---|
| PyTorch | 2.12.0+rocm7.2 |
| HIP | 7.2.53211 |
| GPU | AMD Instinct MI300A (94.2 GB unified memory) |
| OS | RHEL 8, kernel 4.18.0-553.121.1.el8_10 |
| ROCm | 7.2.0 |
| Python | 3.12.7 |

**Not affected**: PyTorch 2.6.0+rocm6.2.4 (same code, same hardware).

---

## Minimal Reproducible Example

```python
"""
SIGSEGV (exit 139) after ~27 iterations of a CTMRG warmup loop.
Crash is silent — no Python exception, process killed by OS.
"""
import torch

DTYPE = torch.float64
DEVICE = torch.device('cuda')
D, chi = 6, 180

def normalize(x):
    return x / torch.linalg.norm(x)

def qr_projector(C, T):
    chi_, D_ = C.shape[0], T.shape[1]
    CT = torch.einsum('ij,jklm->iklm', C, T).reshape(chi_ * D_ * D_, chi_)
    q, r = torch.linalg.qr(CT, mode='reduced')
    return q.reshape(chi_, D_, D_, chi_), r

torch.manual_seed(987654321)
params = torch.rand(2 * D**4, device=DEVICE, dtype=DTYPE) - 0.5
A = params.reshape(2, D, D, D, D)
for p in [(0,1,4,3,2),(0,3,2,1,4),(0,4,1,2,3),(0,3,4,1,2)]:
    A = A + A.permute(p)
M = normalize(A)

with torch.no_grad():
    Mn = normalize(M)
    T_ = torch.einsum('xabcd,xaefg->becfgd', Mn, Mn.conj())
    C_ = torch.einsum('beccgd->begd', T_)
    pad = chi - D**2
    C_ = torch.nn.functional.pad(C_.reshape(D**2, D**2), (0, pad, 0, pad))
    T_ = torch.nn.functional.pad(T_.reshape(D**2, D, D, D**2), (0, pad, 0, 0, 0, 0, 0, pad))
    C_env = normalize(C_.to(device=DEVICE, dtype=DTYPE))
    T_env = normalize(T_.to(device=DEVICE, dtype=DTYPE))

C, T = C_env.clone(), T_env.clone()
with torch.no_grad():
    for i in range(40):
        v, r = qr_projector(C, T)
        T = torch.einsum("ibfj,iaep,xabcd,xefgh,jcgq->pdhq", T, v, M, M.conj(), v)
        C = torch.einsum("ijkl,la,ajkx->ix", T, r, v)
        C = normalize(C + C.T.conj())
        T = normalize(T + T.permute(3, 1, 2, 0).conj())
        print(f"iter {i}: ok", flush=True)
# Expected: prints iter 0..39
# Actual:   prints iter 0..26, then SIGSEGV (exit 139)
```

**Important**: this MWE only crashes when preceded by warmup loops at smaller D (D=2..5 each with 40 steps). The crash threshold is cumulative within a process session. Run the full reproducer:

<details>
<summary>Full reproducer (D=2..5 warmup → D=6 crash)</summary>

```python
import torch, sys
DTYPE = torch.float64
DEVICE = torch.device('cuda')

def normalize(x): return x / torch.linalg.norm(x)
def qr_step(C, T):
    chi, D = C.shape[0], T.shape[1]
    CT = torch.einsum('ij,jklm->iklm', C, T).reshape(chi*D*D, chi)
    q, r = torch.linalg.qr(CT, mode='reduced')
    return q.reshape(chi, D, D, chi), r
def update_T(T, v, M):
    return torch.einsum('ibfj,iaep,xabcd,xefgh,jcgq->pdhq', T, v, M, M.conj(), v)
def sym(A):
    for p in [(0,1,4,3,2),(0,3,2,1,4),(0,4,1,2,3),(0,3,4,1,2)]: A = A + A.permute(p)
    return normalize(A)
def init_env(M, chi):
    D = M.shape[1]
    with torch.no_grad():
        Mn = normalize(M)
        T_ = torch.einsum('xabcd,xaefg->becfgd', Mn, Mn.conj())
        C_ = torch.einsum('beccgd->begd', T_)
        pad = chi - D**2
        C_ = torch.nn.functional.pad(C_.reshape(D**2,D**2),(0,pad,0,pad))
        T_ = torch.nn.functional.pad(T_.reshape(D**2,D,D,D**2),(0,pad,0,0,0,0,0,pad))
    return normalize(C_.to(DEVICE,DTYPE)), normalize(T_.to(DEVICE,DTYPE))

# D=2..5 warmup (each 40 steps) — accumulates allocator state
for D, chi in [(2,20),(3,45),(4,80),(5,125)]:
    M = sym(torch.randn(2,D,D,D,D,device=DEVICE,dtype=DTYPE))
    C, T = init_env(M, chi)
    with torch.no_grad():
        for _ in range(40):
            v, r = qr_step(C, T); T = update_T(T, v, M)
            C = torch.einsum('ijkl,la,ajkx->ix', T, r, v)
            C, T = normalize(C+C.T.conj()), normalize(T+T.permute(3,1,2,0).conj())
    print(f"D={D}: warmup OK")

# D=6 → SIGSEGV at iter ~27
D, chi = 6, 180
M = sym(torch.randn(2,D,D,D,D,device=DEVICE,dtype=DTYPE))
C, T = init_env(M, chi)
with torch.no_grad():
    for i in range(40):
        v, r = qr_step(C, T); T = update_T(T, v, M)
        C = torch.einsum('ijkl,la,ajkx->ix', T, r, v)
        C, T = normalize(C+C.T.conj()), normalize(T+T.permute(3,1,2,0).conj())
        print(f"D=6 iter {i}: ok", flush=True)  # dies at iter ~27
```
</details>

---

## What I've Ruled Out

| Hypothesis | Test | Result |
|---|---|---|
| TunableOp rotating buffer | `PYTORCH_TUNABLEOP_ENABLED=0` | ❌ still crashes |
| hipBLASLt workspace | `TORCH_BLAS_PREFER_HIPBLASLT=0` | ❌ still crashes |
| GC threshold too high | `PYTORCH_CUDA_ALLOC_CONF=garbage_collection_threshold:0.5` | ❌ no effect |
| Block split size | `PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:128` | ❌ no effect |
| Pure C++ hipBLAS | 4-chain DGEMM × 100 iterations (no PyTorch) | ✅ no crash |
| hipTensor | hipTensor contraction × 100 iterations | ✅ no crash |

The last two entries confirm **the bug is in PyTorch's allocator layer, not in ROCm compute kernels**.

---

## Root Cause

`torch.cuda.empty_cache()` (= `release_cached_blocks()`) called every 20 iterations **completely eliminates the crash** — this pins the bug to **accumulated state in `DeviceCachingAllocator`**.

After D=2..5 warmup, the allocator holds ~700 MB reserved with only ~163 MB active. The deferred-free bookkeeping (the `cuda_events` map and cached-block pool) accumulates across alloc/free cycles. When D=6 triggers another large allocation, `hipMalloc` SIGSEGV's instead of returning `hipErrorOutOfMemory` — this is a ROCm 7.x regression (same code is stable on ROCm 6.2.4).

**Why `PYTORCH_CUDA_ALLOC_CONF` options don't help**: `garbage_collection_threshold` and `max_split_size_mb` control behavior *after* `hipMalloc` returns an error. They never fire when `hipMalloc` kills the process via SIGSEGV.

---

## Workarounds (confirmed working)

**Best (zero-overhead on measurement steps)**:
```python
# Add to CTMRG.__call__ warmup loop:
for i in range(warmup):
    ...
    if (i + 1) % 20 == 0:
        torch.cuda.empty_cache()  # ROCm 7.x workaround
```

**Nuclear (fixes everything, ~5× slower)**:
```bash
PYTORCH_NO_HIP_MEMORY_CACHING=1 python your_script.py
```

---

## Proposed Fix in PyTorch

Add a periodic `release_cached_blocks()` call inside `DeviceCachingAllocator::malloc()` under `#if defined(USE_ROCM)`:

**File**: `c10/cuda/CUDACachingAllocator.cpp`  
**Function**: `DeviceCachingAllocator::malloc()`  
**Insert after**: `params.stat_types = get_stat_types_for_pool(pool);`  
**Insert before**: `// First, try to get a block from the existing pool.`

```cpp
#if defined(USE_ROCM)
    // ROCm 7.x workaround: periodic cache release to prevent hipMalloc SIGSEGV.
    //
    // PyTorch's deferred-free bookkeeping accumulates internal event/block state
    // across allocations. After ~27 large allocations in a tight loop the HIP
    // driver's internal state becomes inconsistent and hipMalloc segfaults
    // (SIGSEGV / exit 139) instead of returning hipErrorOutOfMemory.
    //
    // Verified fix (Python-side): torch.cuda.empty_cache() every 20 iterations
    // completely eliminates the crash with <0.3% overhead. torch.cuda.empty_cache()
    // maps to release_cached_blocks(), so we replicate that call here on a fixed
    // period of every 32 malloc calls.
    //
    // PYTORCH_NO_HIP_MEMORY_CACHING=1 also fixes it but is 5x slower.
    // See: https://github.com/pytorch/pytorch/issues/<this issue>
    if (captures_underway.empty()) {
      thread_local int32_t rocm_periodic_flush_count = 0;
      if ((++rocm_periodic_flush_count & 0x1F) == 0) {  // every 32 malloc calls
        release_cached_blocks(context);
      }
    }
#endif // defined(USE_ROCM)
```

**Performance impact**: ~0.3% overhead. `release_cached_blocks` only pays the cost of processing already-completed events and freeing the (usually small) cached pool; when the pool is already empty it is essentially free.

---

## Performance Data (AMD MI300A vs H100, heisenberg.py QR-CTMRG benchmark)

After upgrading from ROCm 6.2.4 → 7.2 + this workaround:

| D | chi | MI300A ROCm 7.2 | H100 | Ratio |
|---|-----|-----------------|------|-------|
| 2 | 20 | 0.028s | 0.032s | **0.88×** (faster) |
| 3 | 45 | 0.037s | 0.037s | **1.00×** |
| 4 | 80 | 0.045s | 0.043s | **1.04×** |
| 5 | 125 | 0.107s | 0.074s | 1.44× |
| 6 | 180 | 0.747s* | 0.320s | 2.34×* |

*D=6 uses Python-side workaround (empty_cache every 20 steps); with the PyTorch patch it should have zero measurement overhead.

---

## cc

@ngimel @malfet @jeffdaily (ROCm allocator)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

**Comment by qiyang-ustc** [comment]
## Update: Corrected MWE and Root Cause Clarification

The original MWE in the issue description was incomplete — it crashes but **only when each D=2..5 warmup is followed by an AD step with `backward()`**. Here is the corrected, verified reproducer:

```python
"""
Verified crash (exit 139, SIGSEGV) on PyTorch 2.12.0+rocm7.2, AMD MI300A.
Requires: D=2..5 warmup + backward() to accumulate cuda_events, then D=6 triggers crash.
"""
import torch
DTYPE = torch.float64
DEVICE = torch.device('cuda')

def normalize(x): return x / torch.linalg.norm(x)

class CTMRG:
    @staticmethod
    def _qr(C, T):
        chi, D = C.shape[0], T.shape[1]
        CT = torch.einsum('ij,jklm->iklm', C, T).reshape(chi*D*D, chi)
        q, r = torch.linalg.qr(CT, mode='reduced')
        return q.reshape(chi, D, D, chi), r
    @staticmethod
    def _update_T(T, v, M):
        return torch.einsum('ibfj,iaep,xabcd,xefgh,jcgq->pdhq', T, v, M, M.conj(), v)
    def __call__(self, M, env, warmup=0, ADiter=0):
        C, T = env[0].detach(), env[1].detach()
        with torch.no_grad():
            for _ in range(warmup):
                v, r = self._qr(C, T); T = self._update_T(T, v, M)
                C = torch.einsum('ijkl,la,ajkx->ix', T, r, v)
                C, T = normalize(C+C.T.conj()), normalize(T+T.permute(3,1,2,0).conj())
        for _ in range(ADiter):   # with autograd graph (M.requires_grad=True)
            v, r = self._qr(C, T); T = self._update_T(T, v, M)
            C = torch.einsum('ijkl,la,ajkx->ix', T, r, v)
            C, T = normalize(C+C.T.conj()), normalize(T+T.permute(3,1,2,0).conj())
        return [C, T]

def sym(A):
    for p in [(0,1,4,3,2),(0,3,2,1,4),(0,4,1,2,3),(0,3,4,1,2)]: A = A + A.permute(p)
    return normalize(A)

def init_env(M, chi):
    D = M.shape[1]
    with torch.no_grad():
        Mn = normalize(M)
        T_ = torch.einsum('xabcd,xaefg->becfgd', Mn, Mn.conj())
        C_ = torch.einsum('beccgd->begd', T_)
        pad = chi - D**2
        C_ = torch.nn.functional.pad(C_.reshape(D**2,D**2),(0,pad,0,pad))
        T_ = torch.nn.functional.pad(T_.reshape(D**2,D,D,D**2),(0,pad,0,0,0,0,0,pad))
    return [normalize(C_.to(DEVICE,DTYPE)), normalize(T_.to(DEVICE,DTYPE))]

ctmrg = CTMRG()
ad_steps = {2:10, 3:12, 4:14, 5:16}

# D=2..5: warmup + AD step + backward() — this is the key trigger
for D, chi in [(2,20),(3,45),(4,80),(5,125)]:
    params = (torch.rand(2*D**4, device=DEVICE, dtype=torch.float64)-0.5).requires_grad_(True)
    M = sym(normalize(params.reshape(2,D,D,D,D).to(DTYPE)))
    env = init_env(M, chi)
    env = ctmrg(M, env, warmup=40, ADiter=0)
    env2 = ctmrg(M, env, warmup=2, ADiter=ad_steps[D])   # creates autograd graph
    loss = sum(x.sum() for x in env2)
    loss.backward()      # frees saved_tensors → floods cuda_events with freed blocks
    torch.cuda.synchronize()
    params.grad = None
    print(f"D={D}: OK", flush=True)

# D=6: crash inside the warmup loop (no iteration count printed — SIGSEGV kills process)
print("D=6: starting warmup — expect SIGSEGV (exit 139)...", flush=True)
D, chi = 6, 180
params = (torch.rand(2*D**4, device=DEVICE, dtype=torch.float64)-0.5).requires_grad_(True)
M = sym(normalize(params.reshape(2,D,D,D,D).to(DTYPE)))
env = init_env(M, chi)
env = ctmrg(M, env, warmup=40, ADiter=0)   # <-- SIGSEGV here
print("PASSED (unexpected!)")
```

**Expected output:**
```
D=2: OK
D=3: OK
D=4: OK
D=5: OK
D=6: starting warmup — expect SIGSEGV (exit 139)...
Segmentation fault (core dumped)
```

---

## Key diagnostic: `torch.cuda.synchronize()` between iterations prevents crash

Adding `torch.cuda.synchronize()` after **each D=6 warmup iteration** makes the crash disappear entirely. This is a strong clue about the mechanism:

- **Without sync**: `process_events()` is called during a D=6 allocation while there are pending events from freed D=2..5 autograd `saved_tensors`. On ROCm 7.2, querying these events causes SIGSEGV.
- **With sync**: all GPU work completes before the next allocation, so all events in `cuda_events` are already done and can be queried safely.

This pinpoints the crash to `DeviceCachingAllocator::process_events()` being called with stale event handles on ROCm 7.2.

`torch.cuda.empty_cache()` (= `release_cached_blocks()` = `synchronize_and_free_events()` + free) also prevents the crash for the same reason: it synchronizes all pending events before freeing, clearing the `cuda_events` map.

---

## Refined patch suggestion

Given that `process_events()` is the crash site, an alternative (possibly more targeted) fix is to call `synchronize_and_free_events()` at the start of `DeviceCachingAllocator::malloc()` on ROCm, rather than periodically calling `release_cached_blocks()`. This is cheaper (only syncs events, doesn't free blocks) while still clearing the hazardous `cuda_events` entries:

```cpp
// In DeviceCachingAllocator::malloc(), before get_free_block(params):
#if defined(USE_ROCM)
  // ROCm 7.x: process_events() can SIGSEGV when querying event handles
  // associated with freed autograd saved_tensors whose underlying streams
  // have changed state. Proactively synchronize pending events to clear
  // the cuda_events map before any allocation that might trigger process_events().
  if (captures_underway.empty()) {
    thread_local int32_t rocm_sync_count = 0;
    if ((++rocm_sync_count & 0x1F) == 0) {  // every 32 malloc calls
      synchronize_and_free_events(context);
    }
  }
#endif
```

Using `synchronize_and_free_events` instead of `release_cached_blocks` avoids unnecessary `hipFree` calls while still preventing the crash.

**Comment by qiyang-ustc** [comment]
Closing — issue needs more investigation before submission. Will reopen once root cause is fully confirmed and MWE is tightened.
